services/src/main/java/org/jboss/windup/web/services/rest/RegisteredApplicationEndpointImpl.java

```diff
@@ -36,23 +47,112 @@
     {
-        String filename = Paths.get(application.getInputPath()).getFileName().toString();
-        Path outputPath = webPathUtil.createWindupReportOutputPath(filename);
-        application.setOutputPath(outputPath.toAbsolutePath().toString());
-        LOG.info("Registering an application at: " + application.getInputPath() + " with output directory: " + outputPath);
+        Map<String, List<InputPart>> uploadForm = data.getFormDataMap();
+        List<InputPart> inputParts = uploadForm.get("file");
 
-        entityManager.persist(application);
+        List<RegisteredApplication> registeredApplications = new ArrayList<>();
 
-        if (application.getApplicationGroup() != null)
-        {
-            if (!application.getApplicationGroup().getApplications().contains(application))
-                application.getApplicationGroup().getApplications().add(application);
+        for (InputPart inputPart : inputParts) {
+            RegisteredApplication application = this.createApplicationFromInputPart(inputPart, appGroupId);
+            registeredApplications.add(application);
+        }
 
-            entityManager.merge(application.getApplicationGroup());
+        return registeredApplications;
+    }
+
+    protected RegisteredApplication createApplicationFromInputPart(InputPart inputPart, long appGroupId) {
+        ApplicationGroup group = this.entityManager.find(ApplicationGroup.class, appGroupId);
+
+        if (group == null) {
+            throw new BadRequestException("Application group not found");
+        }
+
+        try {
+            RegisteredApplication application = new RegisteredApplication();
+
+            MultivaluedMap<String, String> header = inputPart.getHeaders();
+            String fileName = this.getFileName(header);
+
+            //convert the uploaded file to inputstream
+            InputStream inputStream = inputPart.getBody(InputStream.class,null);
+
+            String filePath = Paths.get(this.webPathUtil.getAppPath().toString(), fileName).toString();
+
+            this.saveFileTo(inputStream, filePath);
+
+            application.setInputFilename(fileName);
+            application.setInputPath(filePath);
+
+            group.getApplications().add(application);
+
+            this.entityManager.persist(application);
+            this.entityManager.merge(group);
+
+            return application;
+        } catch (IOException ex) {
+            Logger.getLogger(FileEndpointImpl.class.getName()).log(Level.SEVERE, null, ex);
+            throw new InternalServerErrorException("Error during file upload");
+        }
+    }
+
+
+    protected String getFileName(MultivaluedMap<String, String> header)
+    {
+        String[] contentDisposition = header.getFirst("Content-Disposition").split(";");
+
+        for (String filename : contentDisposition) {
+            if ((filename.trim().startsWith("filename"))) {
```

Comment:
Its not critically important right now, but is it safe to trust the client provided filename in this way? Would it be better to sanitize the input to avoid any ".." or "/" characters that could potentially cause it to store outside of the intended path?